File: loader/src/lib/shell.ts
import { invoke, shell } from '@tauri-apps/api'

export const Shell = {
    /**
     * Expand a folder in file explorer.
     * @param path Absolute path to folder.
     */
    async expandFolder(path: string) {
        await invoke('plugin:shell|expand_folder', {
            path: path
        })
    },

    /**
     * Reveal a file in file explorer.
     * @param path Absolute path to file.
     */
    async revealFile(path: string) {
        await invoke('plugin:shell|reveal_file', {
            path: path
        })
    },

    /**
     * Open an external link.
     * @param url URL.
     */
    async openLink(url: string) {
        if (typeof url === 'string' && url.startsWith('https://')) {
            await shell.open(url)
        }
    },

    /**
     * Extract a zip file to destination directory.
     * @param zipPath Absolute path to zip file.
     * @param destDir Absolute path to destination directory.
     */
    async extractZip(zipPath: string, destDir: string) {
        await invoke('plugin:shell|extract_zip', {
            zipPath,
            destDir
        })
    },

    /**
     * Copy a file to destination path.
     * @param src Absolute path to source file.
     * @param dest Absolute path to destination file.
     */
    async copyFile(src: string, dest: string) {
        await invoke('plugin:shell|copy_file', {
            src,
            dest
        })
    }
}

File: loader/src-tauri/src/shell.rs
use std::process::Command;
use std::fs::File;
use std::io;
use std::path::Path;
use zip::ZipArchive;
use tauri::{
    plugin::{Builder, TauriPlugin},
    Runtime,
};

#[tauri::command]
fn extract_zip(zip_path: &str, dest_dir: &str) -> Result<(), String> {
    let file = File::open(zip_path).map_err(|e| e.to_string())?;
    let mut archive = ZipArchive::new(file).map_err(|e| e.to_string())?;
    
    archive.extract(dest_dir).map_err(|e| e.to_string())?;
    
    Ok(())
}

#[tauri::command]
fn copy_file(src: &str, dest: &str) -> Result<(), String> {
    std::fs::copy(src, dest).map_err(|e| e.to_string())?;
    Ok(())
}

#[tauri::command]
fn expand_folder(path: &str) {
    #[cfg(windows)]
    {
        let path = path.replace("/", "\\");
        Command::new("explorer")
            .args(["/expand,", &path])
            .spawn()
            .unwrap();
    }

    #[cfg(target_os = "macos")]
    Command::new("open").arg(path).spawn().unwrap();
}

#[tauri::command]
fn reveal_file(path: &str) {
    #[cfg(windows)]
    {
        let path = path.replace("/", "\\");
        Command::new("explorer")
            .args(["/select,", &path])
            .spawn()
            .unwrap();
    }

    #[cfg(target_os = "macos")]
    Command::new("open").args(["-R", path]).spawn().unwrap();
}

pub fn init<R: Runtime>() -> TauriPlugin<R> {
    Builder::new("shell")
        .invoke_handler(tauri::generate_handler![expand_folder, reveal_file, extract_zip, copy_file])
        .build()
}
